### DISCUSSION TITLE
Make UI with SVG file

### DISCUSSION THREAD

**Discussion by v479038280** [zauberzeug/nicegui#2973] [Q&A]
Is there an easy way to make the button in the ui clickble, and get input from the SVG file , also change the button state( i.e color ) and the content in the input area
![Untitled Diagram](https://github.com/zauberzeug/nicegui/assets/25096045/1d137817-8da6-4454-a4db-18be7eeeaaec)

**Comment by rodja** [comment]
What do you mean with

> get input from the SVG file

**Reply by v479038280** [reply]
Get value fom the input form,make the button in the picture like a real button which could handle events like click, also change value  in the input area .  
Get element in the SVG file and change its attribute  or get its value

In this way, we can directly use an SVG editor to draw a front-end UI, display it on the interface by using `ui.html('path/to/file.svg')`, and then manipulate the elements within the SVG file to achieve a similar effect as adding interface elements through `ui.element()`.

**Reply by v479038280** [reply]
It seems It's easy to achive such thing with JS thouth I'm not familiar with JS,If is there an easy way with niceGUI

**Reply by snowbollaanm** [reply]
If I understand your question correctly, you can just do sonething like this: 

```py
def some_function():
print('clicked!')

button = ui.html('path/to/file.svg')
button.on('click', some_function)
```

and if you want the cursor to change into a pointer you can add `.style('cursor: pointer;')` to the ui.html

**Reply by v479038280** [reply]
Not just that.
Something like this 
button1=ui.html('path/to/file.svg').success_button
button2=ui.html('path/to/file.svg').secondary_button
button3=ui.html('path/to/file.svg').primary_button
input=ui.html('path/to/file.svg').input
text_area=ui.html('path/to/file.svg').text)_area

handle the element in SVG likeJS,This **make the SVGeditor more like a UIbuilder like Qtdesinger,You can draw the UI first ,Then handle everything in the 'SVG' tag** 

![Untitled Diagram](https://github.com/zauberzeug/nicegui/assets/25096045/ec23d250-062a-48b8-a60e-ee60722c0e99)

**Reply by frankvp11** [reply]
Hi, I think I've figured out the solution to this. In a relatively recent PR I made, we added some events to SVG's, which can be found here. https://nicegui.io/documentation/interactive_image#svg_events
I'm currently working on making a library of sorts that allows you to easily create shapes and whatnot, which can be found here: https://github.com/frankvp11/GraphicMaker/tree/main
I haven't tested it in a while, but it might be a place to start. If you look in the main, you can see some commented-out examples of how it might work. 
I hope that this helps, and if this isn't exactly what you mean, let me know so that I can look further into this because i'm quite interested in this topic.

**Reply by v479038280** [reply]
I am not looking to dynamically draw SVG graphics within an interface using nicegui. Instead, I have an SVG file that has already been created using draw.io. My idea is to manipulate this SVG file to achieve the same functionality as using ui.button() and ui.input(). If this approach is relatively easy to implement, we could design our UI interface in advance with tools like draw.io or Inkscape. Then, we could display it in a browser using ui.html('path/to/file.svg'.readtext()). Subsequently, we would add backend code to extract SVG elements, add or remove their attributes, and change their base property values. For example, in the SVG image, I want to make the button clickable and capable of adding events, and the input box should be able to display the value of a corresponding variable and change dynamically
![Untitled Diagram](https://github.com/zauberzeug/nicegui/assets/25096045/f6cf0dbf-d093-4bf1-a333-7f4f32593585)

**Reply by frankvp11** [reply]
Oh... So you want to parse pre-existing SVG files and be able to 'access' the elements inside it? 
I'll look into that - seems quite interesting

**Reply by v479038280** [reply]
That's what I mean,
This could be a potential approach for visual UI writing with nicegui, similar to qtdesigner. It seems quite easy to change some attributes via JavaScript,like [this](https://www.w3schools.com/graphics/svg_scripting.asp), if we can change any attribute value within the SVG source file in a convenient manner through nicegui, it would mean that Inkscape or draw.io could become a 'Qt Designer' equivalent for nicegui.

**Reply by frankvp11** [reply]
That's actually really interesting. I'll start a repo for this development, I've already gotten quite far with my previous implementation (that I linked above), and am working on the parser right now. 
Edit: Here is the repo:
https://github.com/frankvp11/SVGParser

**Reply by frankvp11** [reply]
I wonder if we would be able to make a UI which nicely displays all the SVG elements so that we can easily modify them - something like a tree down approach, similar to the nicegui json editor but then "nicer" looking?

**Reply by frankvp11** [reply]
Hello! It's me again!
I've got something set up which is honestly quite nice. I encourage you to give it a try, but I've also attached a video here
[Screencast from 2024-04-30 01:17:24 PM.webm](https://github.com/zauberzeug/nicegui/assets/90110502/2e24c5fd-985e-4a08-9f50-92270547151c)
Edit: Now to just make an export function of sorts

**Reply by v479038280** [reply]
I found a library where we can export the svg we need ,
![微信截图_20240509221710](https://github.com/zauberzeug/nicegui/assets/25096045/40960848-2cf9-4767-a290-2c3a2847bcc1)

**Reply by v479038280** [reply]
@frankvp11  I found this [libray  SVG.js ](https://svgjs.dev/docs/3.0/)from [FUXA  ](https://github.com/frangoteam/FUXA?tab=readme-ov-file),if we could import such library  into nicegui then we could manupulate SVG @rodja @falkoschindler

**Reply by frankvp11** [reply]
Seems interesting; but it is slower than raw JS; something that should be considered

**Reply by v479038280** [reply]
ui.run_javascript. would be OK?

**Reply by v479038280** [reply]
@frankvp11 Hope this [library](https://github.com/SVG-Edit/svgedit) is usefull for you 
I made this effect with it 
![微信截图_20240517110100](https://github.com/zauberzeug/nicegui/assets/25096045/86c53912-e443-49ca-9707-b645301bfc4f)

>

![微信截图_20240517110707](https://github.com/zauberzeug/nicegui/assets/25096045/9bc3d60d-9344-403c-b26f-623a7cff8184)

**Reply by v479038280** [reply]
How about ui.svg_editor() by https://github.com/SVG-Edit/svgedit, 
 and.      ui.svg() based on https://svgjs.dev/docs/3.0.    or ui.handle_svg_in_the_page() also based on svgjs.dev @rodja @falkoschindler

**Reply by rodja** [reply]
That are all very complex elements. Maybe this is better suited to be an additional package on PyPI? Similar to https://github.com/zauberzeug/nicegui-highcharts?

**Reply by v479038280** [reply]
I found [this](https://github.com/cduck/drawsvg):A Python 3 library for programmatically generating SVG images and animations that can render and display your drawings in a Jupyter notebook or Jupyter lab. They both could generate and animate. It seems that they both couldn't handle the existing SVG file,I'm still learning ,not sure about that
>update [How can I modify an element that already exists in the dom?](https://svgjs.dev/docs/3.0/faq/#how-can-i-modify-an-element-that-already-exists-in-the-dom)
and [How can I load svg from disk to modify it later?](https://github.com/cduck/drawsvg/discussions/50)

svgjs.dev  could modify the existing file

**Reply by frankvp11** [reply]
I'm not entire sure how [this one](https://github.com/cduck/drawsvg) is different from the package that I created a while ago, minus the fact that it probably won't integrate too well with nicegui. 
Also, I agree with rodja, these are really complex elements to make. I personally agree that they wouldn't fit well with nicegui's main package, and would require excessive amounts of work to make for little use. If you need something specific, I can help, but making an entire feature would be futile.

**Reply by v479038280** [reply]
I agree too,ui.run_javascript() maybe the right way. By the way,I think that Snap.svg is the most suitable tool for handling this kind of task.

**Comment by v479038280** [comment]
> That are all very complex elements. Maybe this is better suited to be an additional package on PyPI? Similar to https://github.com/zauberzeug/nicegui-highcharts?

That's great ,
I'm not sure if snap.svg is better for  manipulating both existing SVG content, and SVG content generated with Snap.
Here is an example [with inkscape and snap](https://goinkscape.com/how-to-create-a-working-clock-in-inkscape/)
and [ANIMATING A MORE ADVANCED ICON WITH SNAP.SVG
](https://goinkscape.com/animating-a-more-advanced-icon-with-snap-svg/
)
